File: OpenSuperWhisper.Core/CrashReporter.cs
namespace OpenSuperWhisper.Core;

/// <summary>
/// F30: writes a sanitized, self-contained crash report file whenever an unhandled exception
/// takes the app down - separate from the rolling log.txt so there's one obvious file to attach
/// when asking for help. Deliberately never touches history.json/settings.json - the exception
/// object and stack trace are the only inputs, so there is no path for dictated speech content to
/// end up in a report. Keeps only the most recent <see cref="MaxReports"/> files so a machine that
/// crashes repeatedly doesn't accumulate reports forever.
/// </summary>
public static class CrashReporter
{
    private const int MaxReports = 10;
    private static readonly string ReportsDir;

    static CrashReporter()
    {
        ReportsDir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "OpenSuperWhisper", "crash-reports");
    }

    /// <summary>Returns the path written to, or null if writing the report itself failed (in
    /// which case the exception is still caught, not rethrown - a crash reporter must never be
    /// the thing that crashes the crash handler).</summary>
    public static string? Write(Exception ex, string context)
    {
        try
        {
            Directory.CreateDirectory(ReportsDir);
            // Timestamp plus a short random suffix (not just milliseconds): a tight crash loop
            // can produce more than one report within the same millisecond, and a collision
            // would silently overwrite the earlier report instead of keeping both.
            var suffix = Guid.NewGuid().ToString("N")[..6];
            var path = Path.Combine(ReportsDir, $"crash-{DateTimeOffset.Now:yyyyMMdd-HHmmss-fff}-{suffix}.txt");

            var text =
                $"OpenSuperWhisper 崩溃报告{Environment.NewLine}" +
                $"时间：{DateTimeOffset.Now:yyyy-MM-dd HH:mm:ss.fff}{Environment.NewLine}" +
                $"版本：{AppVersion.Current}{Environment.NewLine}" +
                $"系统：{Environment.OSVersion}{Environment.NewLine}" +
                $"场景：{context}{Environment.NewLine}" +
                $"{Environment.NewLine}--- 异常详情（不含任何听写内容，只有程序自身的报错信息）---{Environment.NewLine}" +
                FormatException(ex);

            File.WriteAllText(path, text);
            RotateOldReports();
            return path;
        }
        catch (Exception writeEx)
        {
            Log.Error("写入崩溃报告本身失败", writeEx);
            return null;
        }
    }

    private static string FormatException(Exception ex)
    {
        var lines = new List<string>();
        var current = ex;
        var depth = 0;
        while (current != null)
        {
            lines.Add($"[{(depth == 0 ? "异常" : "内层异常")}] {current.GetType().FullName}: {current.Message}");
            lines.Add(current.StackTrace ?? "（无堆栈信息）");
            current = current.InnerException;
            depth++;
        }
        return string.Join(Environment.NewLine + Environment.NewLine, lines);
    }

    private static void RotateOldReports()
    {
        try
        {
            var files = new DirectoryInfo(ReportsDir).GetFiles("crash-*.txt");
            if (files.Length <= MaxReports) return;

            foreach (var f in files.OrderByDescending(f => f.CreationTimeUtc).Skip(MaxReports))
                f.Delete();
        }
        catch (Exception ex)
        {
            Log.Error("清理旧崩溃报告失败", ex);
        }
    }
}

File: src/crash_reporter.rs
//! Panic-time crash reporting: writes a self-contained, privacy-safe crash
//! report to `%LOCALAPPDATA%\Aevocis\crash-reports\` whenever this process
//! panics -- mirrors the shipping C# app's
//! `src-reference/OpenSuperWhisper.Core/CrashReporter.cs`, which writes a
//! dedicated report file (separate from the rolling log) precisely because
//! there should be one obvious file to attach when asking for help.
//!
//! Deliberately never reads `history.json` or `settings.json` -- the panic
//! payload, its source location, and a backtrace are the only inputs, so
//! there is no path for dictated speech content to end up in a report,
//! matching the C# app's explicit design choice.
//!
//! This hook only *observes* the panic: it does not prevent it from
//! propagating/unwinding. This crate's Cargo profile does not set
//! `panic = "abort"`, so Rust's default unwind behavior still runs after
//! this hook returns, which is what lets `Drop` impls -- e.g. tray-icon's
//! `Shell_NotifyIcon(NIM_DELETE)` cleanup -- still fire during unwinding,
//! mirroring the C# app's crash-time tray-icon-cleanup fix.

use std::io;
use std::path::{Path, PathBuf};

use windows::Win32::System::SystemInformation::GetLocalTime;

/// Newest reports beyond this count are deleted on each write -- matches the
/// C# app's `CrashReporter.MaxReports`.
const MAX_REPORTS: usize = 10;

/// Installs the panic hook. Call once, as early as possible in `main()`.
///
/// Chains the previous (default) hook rather than replacing it outright, so
/// the usual "thread panicked at ..." console output is unaffected -- this
/// only adds the crash-report file as a side effect alongside the existing
/// panic diagnostics, it doesn't take them away.
pub fn install() {
    let previous = std::panic::take_hook();
    std::panic::set_hook(Box::new(move |info| {
        // A failing crash reporter must never itself panic (that would
        // abort mid-unwind) or mask the original panic -- any I/O failure
        // here is swallowed and just logged to stderr.
        if let Err(e) = write_report(info) {
            eprintln!("crash_reporter: failed to write crash report: {e}");
        }
        previous(info);
    }));
}

fn reports_dir() -> PathBuf {
    crate::app_data_dir().join("crash-reports")
}

fn write_report(info: &std::panic::PanicHookInfo) -> io::Result<()> {
    let dir = reports_dir();
    std::fs::create_dir_all(&dir)?;

    // GetLocalTime's fields give us a full date+time directly -- no need for
    // an external date/time crate for one timestamp format, consistent with
    // this crate's existing direct-Win32-call style (see history.rs::now_hhmm).
    let st = unsafe { GetLocalTime() };
    let file_stamp = format!(
        "{:04}{:02}{:02}-{:02}{:02}{:02}",
        st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond
    );
    let display_time = format!(
        "{:04}-{:02}-{:02} {:02}:{:02}:{:02}",
        st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond
    );

    // A short suffix (not just the second-granularity timestamp): a tight
    // crash loop can produce more than one report within the same second,
    // and a filename collision would silently overwrite the earlier report
    // instead of keeping both. Sub-second time is enough entropy here; no
    // RNG crate is a dependency of this crate.
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.subsec_nanos())
        .unwrap_or(0);
    let suffix = format!("{:06x}", nanos & 0x00FF_FFFF);

    let path = dir.join(format!("crash-{file_stamp}-{suffix}.txt"));

    let payload = info.payload();
    let message = if let Some(s) = payload.downcast_ref::<&str>() {
        (*s).to_string()
    } else if let Some(s) = payload.downcast_ref::<String>() {
        s.clone()
    } else {
        "<non-string panic payload>".to_string()
    };

    let location = info
        .location()
        .map(|l| format!("{}:{}:{}", l.file(), l.line(), l.column()))
        .unwrap_or_else(|| "<unknown>".to_string());

    // Captured unconditionally rather than gated on RUST_BACKTRACE: a crash
    // report that's useless without an env var the user didn't know to set
    // defeats the point of having a crash report at all.
    let backtrace = std::backtrace::Backtrace::force_capture();

    let text = format!(
        "Aevocis crash report\n\
         Time: {display_time}\n\
         Version: {}\n\
         OS: Windows ({})\n\
         Panic: {message}\n\
         Location: {location}\n\
         Backtrace:\n{backtrace}\n",
        env!("CARGO_PKG_VERSION"),
        std::env::consts::ARCH,
    );

    std::fs::write(&path, text)?;
    rotate(&dir)?;
    Ok(())
}

/// Keeps only the newest `MAX_REPORTS` files in `dir`, deleting older ones.
/// The `crash-{yyyyMMdd-HHmmss}-{suffix}.txt` naming sorts chronologically
/// as plain strings, so a lexicographic sort is sufficient -- no need to
/// stat mtimes.
fn rotate(dir: &Path) -> io::Result<()> {
    let mut files: Vec<PathBuf> = std::fs::read_dir(dir)?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| {
            p.file_name()
                .and_then(|n| n.to_str())
                .is_some_and(|n| n.starts_with("crash-") && n.ends_with(".txt"))
        })
        .collect();
    files.sort();
    if files.len() > MAX_REPORTS {
        for old in &files[..files.len() - MAX_REPORTS] {
            let _ = std::fs::remove_file(old);
        }
    }
    Ok(())
}
